java-checks-test-sources/src/main/files/non-compiling/checks/RedundantTypeCastCheck.java

```diff
@@ -0,0 +1,164 @@
+package checks;
+
+import java.util.function.Predicate;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Comparator;
+import java.util.Set;
+
+class Outer {
+  class A {
+  }
+  class B extends A { }
+  List list;
+  List<String> foo() {
+    Object obj = null;
+    Object o1 = (List<String>) foo(); // Noncompliant [[sc=18;ec=30]] {{Remove this unnecessary cast to "List".}}
+    Object o2 = (List<? extends String>) foo(); // Noncompliant {{Remove this unnecessary cast to "List".}}
+    Object o3 = (List<? super String>) foo(); // Noncompliant {{Remove this unnecessary cast to "List".}}
+    String s1 = (String) obj; // Compliant
+    String s2 = (String) s1; // Noncompliant {{Remove this unnecessary cast to "String".}}
+    A a = (A) new B(); // Noncompliant {{Remove this unnecessary cast to "A".}}
+    A[][] as = (A[][]) new B[1][1]; // Noncompliant {{Remove this unnecessary cast to "A[][]".}}
+    B b = null;
+    fun(b);
+    fun((B) b); // Noncompliant
+    fun((A) b); // Compliant - exception to distinguish the method to call
+    funBParameter((A) b); // Noncompliant
+    List<B> bees = new java.util.ArrayList<B>();
+    List<A> aaas = (List) bees;
```

Comment:
All this code is compiling and is duplicated in the compiling file. I think we should keep only what is not compiling in this file, to avoid duplication.

Reply:
I tried to remove as many duplications as possible but it's kinda hard to make difference between what you need in non-compiling code. I've already removed much from non-compiling file.